services/scheduler/scheduler.go

```diff
@@ -0,0 +1,316 @@
+func (s *Service) Remove(id string) error {
+	s.taskMx.RLock()
+	if cancel, ok := s.tasks[id]; ok {
+		cancel()
+	}
+	s.taskMx.RUnlock()
+
+	s.jobsMx.Lock()
+	defer s.jobsMx.Unlock()
+	err := s.scheduler.RemoveByTag(id)
+	if err != nil {
+		return err
+	}
+
+	if err := models.RemoveScheduledTask(s.db.Querier, id); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reload removes job from scheduler and add it again from DB.
+func (s *Service) Reload(id string) error {
+	dbTask, err := models.FindScheduledTaskByID(s.db.Querier, id)
+	if err != nil {
+		return err
+	}
+
+	if dbTask.Running {
+		return fmt.Errorf("task is running")
+	}
+
+	s.jobsMx.Lock()
+	defer s.jobsMx.Unlock()
+	var job *gocron.Job
+	for _, j := range s.scheduler.Jobs() {
+		if len(j.Tags()) > 0 && j.Tags()[0] == id {
+			job = j
+			break
+		}
+	}
+
+	if job == nil {
+		return fmt.Errorf("job not found")
+	}
+
+	task, err := s.convertDBTask(dbTask)
+	if err != nil {
+		return err
+	}
+
+	s.scheduler.RemoveByReference(job)
+
+	j := s.scheduler.Cron(dbTask.CronExpression).SingletonMode()
+	if !dbTask.StartAt.IsZero() {
+		j = j.StartAt(dbTask.StartAt)
+	}
+
+	fn := s.wrapTask(task, dbTask.ID, int(dbTask.RetriesRemaining), dbTask.RetryInterval)
+	if _, err := j.Tag(dbTask.ID).Do(fn); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) loadFromDB() error {
+	s.jobsMx.Lock()
+	defer s.jobsMx.Unlock()
+
+	disabled := false
+	dbTasks, err := models.FindScheduledTasks(s.db.Querier, models.ScheduledTasksFilter{
+		Disabled: &disabled,
+	})
+	if err != nil {
+		return err
+	}
+
+	tasks := make([]Task, 0, len(dbTasks))
+	for _, dbTask := range dbTasks {
+		task, err := s.convertDBTask(dbTask)
+		if err != nil {
+			return err
+		}
+		tasks = append(tasks, task)
+	}
+
+	s.scheduler.Clear()
+	// Reset tags
+	s.scheduler.TagsUnique()
+	for i, task := range tasks {
+		dbTask := dbTasks[i]
+		fn := s.wrapTask(task, dbTask.ID, int(dbTask.RetriesRemaining), dbTask.RetryInterval)
+		j := s.scheduler.Cron(dbTask.CronExpression).SingletonMode()
+		if !dbTask.StartAt.IsZero() {
+			j = j.StartAt(dbTask.StartAt)
+		}
+		if _, err := j.Tag(dbTask.ID).Do(fn); err != nil {
+			return err
+		}
+
+	}
+	return nil
+}
+func (s *Service) wrapTask(task Task, id string, retry int, retryInterval time.Duration) func() {
+	return func() {
+		var err error
+		l := s.l.WithFields(logrus.Fields{
+			"id":       id,
+			"taskType": task.Type(),
+		})
+		ctx, cancel := context.WithCancel(context.Background())
+
+		s.taskMx.Lock()
+		s.tasks[id] = cancel
+		s.taskMx.Unlock()
+
+		defer func() {
+			cancel()
+			s.taskMx.Lock()
+			delete(s.tasks, id)
+			s.taskMx.Unlock()
+		}()
+		retriesRemaining := retry
+		succeeded := false
+		for {
+			t := time.Now()
+			l.Debug("Starting task")
+			_, err = models.ChangeScheduledTask(s.db.Querier, id, models.ChangeScheduledTaskParams{
+				Running: pointer.ToBool(true),
+			})
+
+			err = task.Do(ctx)
```

Comment:
as I see in the case of `retriesRemaining` == n you will perform n+1 actual attempts 😄

Reply:
let's also cover this case in the test

Reply:
job should run at least one time.
If retries == 0,  it will run just one time.
If retries == 1, it will run max two times. One run, and one retry.

Also I don't know how to test this. Any idea? I can schedule running job with minimum frequency 1min. And then this job would be long.

Reply:
I got the idea, then it's ok. 

Regarding the testing, you could set `retryInterval` to 1ms and have a mock for Do() function, which will return an error couple of times and a successful last call. On the mock's side expect all these calls. Also, you could check `retriesRemaining` in the DB.

Reply:
But still. next run can take up to 60s, before `Do()` function is run.